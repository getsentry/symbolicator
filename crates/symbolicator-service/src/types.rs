//! Types for the Symbolicator API.
//!
//! This module contains some types which (de)serialise to/from JSON to make up the public
//! HTTP API.  Its messy and things probably need a better place and different way to signal
//! they are part of the public API.

use std::collections::BTreeMap;
use std::fmt;
use std::sync::Arc;

use serde::{Deserialize, Serialize};
use symbolicator_sources::ObjectType;

use crate::utils::hex::HexValue;

/// The scope of a source or debug file.
///
/// Based on scopes, access to debug files that have been cached is determined. If a file comes from
/// a public source, it can be used for any symbolication request. Otherwise, the symbolication
/// request must match the scope of a file.
#[derive(Debug, Clone, Deserialize, Serialize, Eq, Ord, PartialEq, PartialOrd, Hash)]
#[serde(untagged)]
#[derive(Default)]
pub enum Scope {
    #[serde(rename = "global")]
    #[default]
    Global,
    Scoped(Arc<str>),
}

impl AsRef<str> for Scope {
    fn as_ref(&self) -> &str {
        match *self {
            Scope::Global => "global",
            Scope::Scoped(ref s) => s,
        }
    }
}

impl fmt::Display for Scope {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match *self {
            Scope::Global => f.write_str("global"),
            Scope::Scoped(ref scope) => f.write_str(scope),
        }
    }
}

/// Configuration for scraping of JS Sources, Source Maps and Source Context from the web.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ScrapingConfig {
    /// Whether scraping should happen at all.
    pub enabled: bool,
    /// Whether Symbolicator should verify SSL certs when scraping from the web.
    ///
    /// Defaults to `true`, just to be safe.
    #[serde(default = "default_verify_ssl")]
    pub verify_ssl: bool,
    /// A list of "allowed origin patterns" that control:
    /// - for sourcemaps: what URLs we are allowed to scrape from.
    /// - for source context: which URLs should be authenticated using attached headers
    ///
    /// Allowed origins may be defined in several ways:
    /// - `http://domain.com[:port]`: Exact match for base URI (must include port).
    /// - `*`: Allow any domain.
    /// - `*.domain.com`: Matches domain.com and all subdomains, on any port.
    /// - `domain.com`: Matches domain.com on any port.
    /// - `*:port`: Wildcard on hostname, but explicit match on port.
    pub allowed_origins: Vec<String>,
    /// A map of headers to send with every HTTP request while scraping.
    pub headers: BTreeMap<String, String>,
}

impl Default for ScrapingConfig {
    fn default() -> Self {
        Self {
            enabled: true,
            // verify_ssl: false,
            allowed_origins: vec!["*".to_string()],
            headers: Default::default(),
            verify_ssl: true,
        }
    }
}

fn default_verify_ssl() -> bool {
    true
}

/// Specification of a module loaded into the process.
#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
pub struct RawObjectInfo {
    /// Platform image file type (container format).
    #[serde(rename = "type")]
    pub ty: ObjectType,

    /// Identifier of the code file.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub code_id: Option<String>,

    /// Name of the code file.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub code_file: Option<String>,

    /// Identifier of the debug file.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub debug_id: Option<String>,

    /// Name of the debug file.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub debug_file: Option<String>,

    /// Checksum of the file's contents.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub debug_checksum: Option<String>,

    /// Absolute address at which the image was mounted into virtual memory.
    ///
    /// We do allow the `image_addr` to be skipped if it is zero. This is because systems like WASM
    /// do not require modules to be mounted at a specific absolute address. Per definition, a
    /// module mounted at `0` does not support absolute addressing.
    #[serde(default)]
    pub image_addr: HexValue,

    /// Size of the image in virtual memory.
    ///
    /// The size is infered from the module list if not specified.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub image_size: Option<u64>,
}

/// Information on a debug information file.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, Eq, PartialEq)]
#[serde(rename_all = "snake_case")]
#[derive(Default)]
pub enum ObjectFileStatus {
    /// The file was found and successfully processed.
    Found,
    /// The image was not referenced in the stack trace and not further handled.
    #[default]
    Unused,
    /// The file could not be found in any of the specified sources.
    Missing,
    /// The file failed to process.
    Malformed,
    /// The file could not be downloaded.
    FetchingFailed,
    /// Downloading or processing the file took too long.
    Timeout,
    /// An internal error while handling this image.
    Other,
}

impl ObjectFileStatus {
    pub fn name(self) -> &'static str {
        // used for metrics
        match self {
            ObjectFileStatus::Found => "found",
            ObjectFileStatus::Unused => "unused",
            ObjectFileStatus::Missing => "missing",
            ObjectFileStatus::Malformed => "malformed",
            ObjectFileStatus::FetchingFailed => "fetching_failed",
            ObjectFileStatus::Timeout => "timeout",
            ObjectFileStatus::Other => "other",
        }
    }
}
