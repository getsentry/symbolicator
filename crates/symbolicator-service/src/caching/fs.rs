use std::io;
use std::path::{Path, PathBuf};
use std::sync::atomic::AtomicIsize;
use std::sync::Arc;
use std::time::{Duration, Instant, SystemTime};

use filetime::FileTime;
use symbolic::common::ByteView;
use tempfile::NamedTempFile;

use crate::config::{CacheConfig, Config};

use super::cache_error::cache_entry_from_bytes;
use super::{CacheEntry, CacheError, CacheName};

/// The interval in which positive caches should be touched.
///
/// Positive "good" caches use a "time to idle" instead of "time to live" mode.
/// We thus need to regularly "touch" the files to signal that they are still in use.
/// This is being debounced to once every hour to not have to touch them on every single use.
const TOUCH_EVERY: Duration = Duration::from_secs(3600);

/// Common cache configuration.
///
/// Many parts of Symbolicator use a cache to save having to re-download data or reprocess
/// downloaded data.  All caches behave similarly and their behaviour is determined by this
/// struct.
#[derive(Debug, Clone)]
pub struct Cache {
    /// Cache identifier used for metric names.
    pub(super) name: CacheName,

    /// Directory to use for storing cache items. Will be created if it does not exist.
    ///
    /// Leaving this as None will disable this cache.
    pub(super) cache_dir: Option<PathBuf>,

    /// Directory to use for temporary files.
    ///
    /// When writing a new file into the cache it is best to write it to a temporary file in
    /// a sibling directory, once fully written it can then be atomically moved to the
    /// actual location withing the [`cache_dir`](Self::cache_dir).
    ///
    /// Just like for `cache_dir` when this cache is disabled this will be `None`.
    tmp_dir: Option<PathBuf>,

    /// Time when this process started.
    start_time: SystemTime,

    /// Options intended to be user-configurable.
    cache_config: CacheConfig,

    /// The maximum number of lazy refreshes of this cache.
    max_lazy_refreshes: Arc<AtomicIsize>,

    /// The capacity (in bytes) of the in-memory cache.
    pub(super) in_memory_capacity: u64,
}

impl Cache {
    pub fn from_config(
        name: CacheName,
        config: &Config,
        cache_config: CacheConfig,
        max_lazy_refreshes: Arc<AtomicIsize>,
        in_memory_capacity: u64,
    ) -> io::Result<Self> {
        let tmp_dir = config.cache_dir("tmp");
        let cache_dir = config.cache_dir(name.as_ref());

        if let Some(ref dir) = cache_dir {
            std::fs::create_dir_all(dir)?;
        }

        Ok(Cache {
            name,
            cache_dir,
            tmp_dir,
            start_time: SystemTime::now(),
            cache_config,
            max_lazy_refreshes,
            in_memory_capacity,
        })
    }

    pub fn name(&self) -> CacheName {
        self.name
    }

    pub fn cache_dir(&self) -> Option<&Path> {
        self.cache_dir.as_deref()
    }

    pub fn max_lazy_refreshes(&self) -> Arc<AtomicIsize> {
        self.max_lazy_refreshes.clone()
    }

    /// Validate cache expiration of path.
    ///
    /// If cache should not be used, `Err(io::ErrorKind::NotFound)` is returned.
    /// If cache is usable, `Ok(x)` is returned with the opened [`ByteView`], and
    /// an [`ExpirationTime`] that indicates whether the file should be touched before using.
    pub(super) fn check_expiry(
        &self,
        path: &Path,
    ) -> io::Result<(CacheEntry<ByteView<'static>>, ExpirationTime)> {
        // We use `mtime` to keep track of both "cache last used" and "cache created" depending on
        // whether the file is a negative cache item or not, because literally every other
        // filesystem attribute is unreliable.
        //
        // * creation time does not exist pre-Linux 4.11
        // * most filesystems are mounted with noatime
        //
        // States a cache item can be in:
        // * negative/empty: An empty file. Represents a failed download. mtime is used to indicate
        //   when the failed download happened (when the file was created)
        // * malformed: A file with the content `b"malformed"`. Represents a failed symcache
        //   conversion. mtime indicates when we attempted to convert.
        // * ok (don't really have a name): File has any other content, mtime is used to keep track
        //   of last use.
        let metadata = path.metadata()?;
        tracing::trace!("File length: {}", metadata.len());

        let bv = ByteView::open(path)?;
        let mtime = metadata.modified()?;
        let mtime_elapsed = mtime.elapsed().unwrap_or_default();

        let cache_entry = cache_entry_from_bytes(bv);
        let expiration_time = match expiration_strategy(&cache_entry) {
            ExpirationStrategy::None => {
                let max_unused_for = self.cache_config.max_unused_for().unwrap_or(Duration::MAX);

                if mtime_elapsed > max_unused_for {
                    return Err(io::ErrorKind::NotFound.into());
                }

                // we want to touch good caches once every `TOUCH_EVERY`
                let touch_in = TOUCH_EVERY.saturating_sub(mtime_elapsed);
                ExpirationTime::TouchIn(touch_in)
            }
            ExpirationStrategy::Negative => {
                let retry_misses_after = self
                    .cache_config
                    .retry_misses_after()
                    .unwrap_or(Duration::MAX);

                let expires_in = retry_misses_after.saturating_sub(mtime_elapsed);

                if expires_in == Duration::ZERO {
                    return Err(io::ErrorKind::NotFound.into());
                }

                ExpirationTime::RefreshIn(expires_in)
            }
            ExpirationStrategy::Malformed => {
                let retry_malformed_after = self
                    .cache_config
                    .retry_malformed_after()
                    .unwrap_or(Duration::MAX);

                let expires_in = retry_malformed_after.saturating_sub(mtime_elapsed);

                // Immediately expire malformed items that have been created before this process started.
                // See docstring of MALFORMED_MARKER
                if mtime < self.start_time || expires_in == Duration::ZERO {
                    tracing::trace!("Created at is older than start time");
                    return Err(io::ErrorKind::NotFound.into());
                }

                ExpirationTime::RefreshIn(expires_in)
            }
        };

        Ok((cache_entry, expiration_time))
    }

    /// Validates `cachefile` against expiration config and open a [`ByteView`] on it.
    ///
    /// Takes care of bumping `mtime`.
    ///
    /// If an open [`ByteView`] is returned it also returns whether the mtime has been
    /// bumped.
    pub fn open_cachefile(
        &self,
        path: &Path,
    ) -> io::Result<Option<(CacheEntry<ByteView<'static>>, ExpirationTime)>> {
        // `io::ErrorKind::NotFound` can be returned from multiple locations in this function. All
        // of those can indicate a cache miss as cache cleanup can run inbetween. Only when we have
        // an open ByteView we can be sure to have a cache hit.
        catch_not_found(|| {
            let (cache_entry, mut expiration) = self.check_expiry(path)?;

            let should_touch = matches!(expiration, ExpirationTime::TouchIn(Duration::ZERO));
            if should_touch {
                filetime::set_file_mtime(path, FileTime::now())?;
                // well, we just touched the file ;-)
                expiration = ExpirationTime::TouchIn(TOUCH_EVERY);
            }

            Ok((cache_entry, expiration))
        })
    }

    /// Create a new temporary file to use in the cache.
    pub fn tempfile(&self) -> io::Result<NamedTempFile> {
        match self.tmp_dir {
            Some(ref path) => {
                // The `cleanup` process could potentially remove the parent directories we are
                // operating in, so be defensive here and retry the fs operations.
                const MAX_RETRIES: usize = 2;
                let mut retries = 0;
                loop {
                    retries += 1;

                    if let Err(e) = std::fs::create_dir_all(path) {
                        sentry::with_scope(
                            |scope| scope.set_extra("path", path.display().to_string().into()),
                            || tracing::error!("Failed to create cache directory: {:?}", e),
                        );
                        if retries > MAX_RETRIES {
                            return Err(e);
                        }
                        continue;
                    }

                    match tempfile::Builder::new().prefix("tmp").tempfile_in(path) {
                        Ok(temp_file) => return Ok(temp_file),
                        Err(e) => {
                            sentry::with_scope(
                                |scope| scope.set_extra("path", path.display().to_string().into()),
                                || tracing::error!("Failed to create cache file: {:?}", e),
                            );
                            if retries > MAX_RETRIES {
                                return Err(e);
                            }
                            continue;
                        }
                    }
                }
            }
            None => Ok(NamedTempFile::new()?),
        }
    }
}

/// Expiration strategies for cache items. These aren't named after the strategies themselves right
/// now but after the type of cache entry they should be used on instead.
#[derive(Debug, PartialEq, Eq)]
pub enum ExpirationStrategy {
    /// Clean up after it is untouched for a fixed period of time.
    None,
    /// Clean up after a forced cool-off period so it can be re-downloaded.
    Negative,
    /// Clean up after it is untouched for a fixed period of time. Immediately clean up if the item
    /// was last touched before the process executing cleanup started.
    Malformed,
}

/// This gives the time at which different cache items need to be refreshed or touched.
#[derive(Debug, Clone)]
pub enum ExpirationTime {
    /// The [`Duration`] after which [`Negative`](ExpirationStrategy::Negative) or
    /// [`Malformed`](ExpirationStrategy::Malformed) cache entries expire and need
    /// to be refreshed.
    RefreshIn(Duration),

    /// The [`Duration`] after which a positive cache entry needs to be touched to keep it
    /// alive for a longer time.
    TouchIn(Duration),
}

impl ExpirationTime {
    /// Gives the [`ExpirationTime`] for a freshly created cache with the given [`CacheEntry`].
    pub fn for_fresh_status<T>(cache: &Cache, entry: &CacheEntry<T>) -> Self {
        let config = &cache.cache_config;
        let strategy = expiration_strategy(entry);
        match strategy {
            ExpirationStrategy::None => {
                // we want to touch good caches once every hour
                Self::TouchIn(Duration::from_secs(3600))
            }
            ExpirationStrategy::Negative => {
                let retry_misses_after = config.retry_misses_after().unwrap_or(Duration::MAX);

                Self::RefreshIn(retry_misses_after)
            }
            ExpirationStrategy::Malformed => {
                let retry_malformed_after = config.retry_malformed_after().unwrap_or(Duration::MAX);

                Self::RefreshIn(retry_malformed_after)
            }
        }
    }

    /// Says whether the cache was just touched.
    pub fn was_touched(&self) -> bool {
        matches!(self, ExpirationTime::TouchIn(TOUCH_EVERY))
    }

    /// Gives the [`Instant`] at which the item expires.
    pub fn as_instant(&self) -> Instant {
        let duration = match self {
            ExpirationTime::RefreshIn(d) => d,
            ExpirationTime::TouchIn(d) => d,
        };
        Instant::now() + *duration
    }
}

/// Checks the cache contents in `buf` and returns the cleanup strategy that should be used
/// for the item.
pub(super) fn expiration_strategy<T>(status: &CacheEntry<T>) -> ExpirationStrategy {
    match status {
        Ok(_) => ExpirationStrategy::None,
        Err(CacheError::Malformed(_)) => ExpirationStrategy::Malformed,
        // All other errors should be treated as "negative" in terms of the expiration.
        Err(_) => ExpirationStrategy::Negative,
    }
}

pub(super) fn catch_not_found<F, R>(f: F) -> io::Result<Option<R>>
where
    F: FnOnce() -> io::Result<R>,
{
    match f() {
        Ok(x) => Ok(Some(x)),
        Err(e) => match e.kind() {
            io::ErrorKind::NotFound => Ok(None),
            _ => Err(e),
        },
    }
}
