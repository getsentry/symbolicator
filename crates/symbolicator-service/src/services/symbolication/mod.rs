use symbolicator_sources::ObjectId;

use crate::services::cficaches::{CfiCacheActor, CfiCacheError};
use crate::services::objects::ObjectsActor;
use crate::services::ppdb_caches::{PortablePdbCacheActor, PortablePdbCacheError};
use crate::services::symcaches::{SymCacheActor, SymCacheError};
use crate::types::{ObjectFileStatus, RawObjectInfo};

mod apple;
mod module_lookup;
mod process_minidump;
// we should really rename this here to the `SymbolicatorService`, as it does a lot more
// than just symbolication ;-)
#[allow(clippy::module_inception)]
mod symbolication;

pub use symbolication::{StacktraceOrigin, SymbolicateStacktraces};

#[derive(Clone, Debug)]
pub struct SymbolicationActor {
    objects: ObjectsActor,
    symcaches: SymCacheActor,
    cficaches: CfiCacheActor,
    ppdb_caches: PortablePdbCacheActor,
    diagnostics_cache: crate::cache::Cache,
}

impl SymbolicationActor {
    pub fn new(
        objects: ObjectsActor,
        symcaches: SymCacheActor,
        cficaches: CfiCacheActor,
        ppdb_caches: PortablePdbCacheActor,
        diagnostics_cache: crate::cache::Cache,
    ) -> Self {
        SymbolicationActor {
            objects,
            symcaches,
            cficaches,
            ppdb_caches,
            diagnostics_cache,
        }
    }
}

impl From<&CfiCacheError> for ObjectFileStatus {
    fn from(e: &CfiCacheError) -> ObjectFileStatus {
        match e {
            CfiCacheError::Fetching(_) => ObjectFileStatus::FetchingFailed,
            // nb: Timeouts during download are also caught by Fetching
            CfiCacheError::Timeout => ObjectFileStatus::Timeout,
            CfiCacheError::ObjectParsing(_) => ObjectFileStatus::Malformed,

            _ => {
                // Just in case we didn't handle an error properly,
                // capture it here. If an error was captured with
                // `capture_error` further down in the callstack, it
                // should be explicitly handled here as a
                // SymCacheError variant.
                sentry::capture_error(e);
                ObjectFileStatus::Other
            }
        }
    }
}

impl From<&SymCacheError> for ObjectFileStatus {
    fn from(e: &SymCacheError) -> ObjectFileStatus {
        match e {
            SymCacheError::Fetching(_) => ObjectFileStatus::FetchingFailed,
            // nb: Timeouts during download are also caught by Fetching
            SymCacheError::Timeout => ObjectFileStatus::Timeout,
            SymCacheError::Malformed => ObjectFileStatus::Malformed,
            SymCacheError::ObjectParsing(_) => ObjectFileStatus::Malformed,
            _ => {
                // Just in case we didn't handle an error properly,
                // capture it here. If an error was captured with
                // `capture_error` further down in the callstack, it
                // should be explicitly handled here as a
                // SymCacheError variant.
                sentry::capture_error(e);
                ObjectFileStatus::Other
            }
        }
    }
}

impl From<&PortablePdbCacheError> for ObjectFileStatus {
    fn from(e: &PortablePdbCacheError) -> ObjectFileStatus {
        match e {
            PortablePdbCacheError::Fetching(_) => ObjectFileStatus::FetchingFailed,
            // nb: Timeouts during download are also caught by Fetching
            PortablePdbCacheError::Timeout => ObjectFileStatus::Timeout,
            PortablePdbCacheError::Malformed => ObjectFileStatus::Malformed,
            PortablePdbCacheError::PortablePdbParsing(_) => ObjectFileStatus::Malformed,
            _ => {
                // Just in case we didn't handle an error properly,
                // capture it here. If an error was captured with
                // `capture_error` further down in the callstack, it
                // should be explicitly handled here as a
                // PortablePdbCacheError variant.
                sentry::capture_error(e);
                ObjectFileStatus::Other
            }
        }
    }
}

fn object_id_from_object_info(object_info: &RawObjectInfo) -> ObjectId {
    ObjectId {
        debug_id: match object_info.debug_id.as_deref() {
            None | Some("") => None,
            Some(string) => string.parse().ok(),
        },
        code_id: match object_info.code_id.as_deref() {
            None | Some("") => None,
            Some(string) => string.parse().ok(),
        },
        debug_file: object_info.debug_file.clone(),
        code_file: object_info.code_file.clone(),
        object_type: object_info.ty,
    }
}
