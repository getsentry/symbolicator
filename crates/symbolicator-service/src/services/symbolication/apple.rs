use std::fs::File;
use std::sync::Arc;

use anyhow::{Context, Result};
use apple_crash_report_parser::AppleCrashReport;
use chrono::{DateTime, Utc};
use regex::Regex;

use symbolic::common::{Arch, CodeId, DebugId};
use symbolicator_sources::{ObjectType, SourceConfig};

use crate::types::{
    CompleteObjectInfo, CompletedSymbolicationResponse, RawFrame, RawObjectInfo, RawStacktrace,
    Scope, SystemInfo,
};
use crate::utils::hex::HexValue;

use super::{StacktraceOrigin, SymbolicateStacktraces, SymbolicationActor};

impl SymbolicationActor {
    #[tracing::instrument(skip_all)]
    fn parse_apple_crash_report(
        &self,
        scope: Scope,
        report: File,
        sources: Arc<[SourceConfig]>,
    ) -> Result<(SymbolicateStacktraces, AppleCrashReportState)> {
        let report =
            AppleCrashReport::from_reader(report).context("failed to parse apple crash report")?;
        let mut metadata = report.metadata;

        let arch = report
            .code_type
            .as_ref()
            .and_then(|code_type| code_type.split(' ').next())
            .and_then(|word| word.parse().ok())
            .unwrap_or_default();

        let modules = report
            .binary_images
            .into_iter()
            .map(map_apple_binary_image)
            .collect();

        let mut stacktraces = Vec::with_capacity(report.threads.len());

        for thread in report.threads {
            let registers = thread
                .registers
                .unwrap_or_default()
                .into_iter()
                .map(|(name, addr)| (name, HexValue(addr.0)))
                .collect();

            let frames = thread
                .frames
                .into_iter()
                .map(|frame| RawFrame {
                    instruction_addr: HexValue(frame.instruction_addr.0),
                    package: frame.module,
                    ..RawFrame::default()
                })
                .collect();

            stacktraces.push(RawStacktrace {
                thread_id: Some(thread.id),
                thread_name: thread.name,
                is_requesting: Some(thread.crashed),
                registers,
                frames,
            });
        }

        let request = SymbolicateStacktraces {
            modules,
            scope,
            sources,
            origin: StacktraceOrigin::AppleCrashReport,
            signal: None,
            stacktraces,
            apply_source_context: true,
            scraping: Default::default(),
        };

        let mut system_info = SystemInfo {
            os_name: metadata.remove("OS Version").unwrap_or_default(),
            device_model: metadata.remove("Hardware Model").unwrap_or_default(),
            cpu_arch: arch,
            ..SystemInfo::default()
        };

        if let Some(captures) = OS_MACOS_REGEX.captures(&system_info.os_name) {
            system_info.os_version = captures
                .name("version")
                .map(|m| m.as_str().to_string())
                .unwrap_or_default();
            system_info.os_build = captures
                .name("build")
                .map(|m| m.as_str().to_string())
                .unwrap_or_default();
            system_info.os_name = "macOS".to_string();
        }

        // https://developer.apple.com/library/archive/technotes/tn2151/_index.html
        let crash_reason = metadata.remove("Exception Type");
        let crash_details = report
            .application_specific_information
            .or_else(|| metadata.remove("Exception Message"))
            .or_else(|| metadata.remove("Exception Subtype"))
            .or_else(|| metadata.remove("Exception Codes"));

        let state = AppleCrashReportState {
            timestamp: report.timestamp,
            system_info,
            crash_reason,
            crash_details,
        };

        Ok((request, state))
    }

    pub async fn process_apple_crash_report(
        &self,
        scope: Scope,
        report: File,
        sources: Arc<[SourceConfig]>,
    ) -> Result<CompletedSymbolicationResponse> {
        let (request, state) = self.parse_apple_crash_report(scope, report, sources)?;
        let mut response = self.symbolicate(request).await?;

        state.merge_into(&mut response);
        Ok(response)
    }
}

lazy_static::lazy_static! {
    /// Format sent by Unreal Engine on macOS
    static ref OS_MACOS_REGEX: Regex = Regex::new(r"^Mac OS X (?P<version>\d+\.\d+\.\d+)( \((?P<build>[a-fA-F0-9]+)\))?$").unwrap();
}

#[derive(Debug)]
struct AppleCrashReportState {
    timestamp: Option<DateTime<Utc>>,
    system_info: SystemInfo,
    crash_reason: Option<String>,
    crash_details: Option<String>,
}

impl AppleCrashReportState {
    fn merge_into(mut self, response: &mut CompletedSymbolicationResponse) {
        if self.system_info.cpu_arch == Arch::Unknown {
            self.system_info.cpu_arch = response
                .modules
                .iter()
                .map(|object| object.arch)
                .find(|arch| *arch != Arch::Unknown)
                .unwrap_or_default();
        }

        response.timestamp = self.timestamp;
        response.system_info = Some(self.system_info);
        response.crash_reason = self.crash_reason;
        response.crash_details = self.crash_details;
        response.crashed = Some(true);
    }
}

fn map_apple_binary_image(image: apple_crash_report_parser::BinaryImage) -> CompleteObjectInfo {
    let code_id = CodeId::from_binary(&image.uuid.as_bytes()[..]);
    let debug_id = DebugId::from_uuid(image.uuid);

    let raw_info = RawObjectInfo {
        ty: ObjectType::Macho,
        code_id: Some(code_id.to_string()),
        code_file: Some(image.path.clone()),
        debug_id: Some(debug_id.to_string()),
        debug_file: Some(image.path),
        debug_checksum: None,
        image_addr: HexValue(image.addr.0),
        image_size: match image.size {
            0 => None,
            size => Some(size),
        },
    };

    raw_info.into()
}
