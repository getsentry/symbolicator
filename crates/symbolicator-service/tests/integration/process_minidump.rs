use std::fs;
use std::io::Write;
use std::sync::Arc;

use tempfile::NamedTempFile;

use symbolicator_service::types::Scope;
use symbolicator_test as test;
use symbolicator_test::assert_snapshot;

use crate::symbolication::setup_service;

macro_rules! stackwalk_minidump {
    ($path:expr) => {
        async {
            let (symbolication, cache_dir) = setup_service(|_| ()).await;
            let (_symsrv, source) = test::symbol_server();

            let minidump = test::read_fixture($path);
            let mut minidump_file = NamedTempFile::new().unwrap();
            minidump_file.write_all(&minidump).unwrap();
            let response = symbolication
                .process_minidump(
                    Scope::Global,
                    minidump_file.into_temp_path(),
                    Arc::new([source]),
                )
                .await;

            assert_snapshot!(response.unwrap());

            let global_dir = cache_dir.path().join("object_meta/global");
            let mut cache_entries: Vec<_> = fs::read_dir(global_dir)
                .unwrap()
                .map(|x| x.unwrap().file_name().into_string().unwrap())
                .collect();

            cache_entries.sort();
            assert_snapshot!(cache_entries);
        }
    };
}

#[tokio::test]
async fn test_minidump_windows() {
    stackwalk_minidump!("windows.dmp").await
}

#[tokio::test]
async fn test_minidump_macos() {
    stackwalk_minidump!("macos.dmp").await
}

#[tokio::test]
async fn test_minidump_linux() {
    stackwalk_minidump!("linux.dmp").await
}
