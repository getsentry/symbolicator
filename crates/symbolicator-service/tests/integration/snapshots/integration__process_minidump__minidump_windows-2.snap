---
source: crates/symbolicator-service/tests/integration/process_minidump.rs
assertion_line: 46
expression: cache_entries
---
- local_59_0285e9e0000_debug
- local_59_b0d8f3183000_debug
- local_5a_b380779000_debug
- local_crash_exe_5AB380779000_crash_ex_
- local_crash_exe_5AB380779000_crash_exe
- local_crash_pdb_3249D99D0C4049318610F4E4FB0B69361_crash_pd_
- local_crash_pdb_3249D99D0C4049318610F4E4FB0B69361_crash_pdb
- local_crash_pdb_3249D99D0C4049318610F4E4FB0B69361_crash_src_zip
- local_crash_pdb_3249D99D0C4049318610F4E4FB0B69361_crash_sym
- local_kernel32_dll_590285E9e0000_kernel32_dl_
- local_kernel32_dll_590285E9e0000_kernel32_dll
- local_ntdll_dll_59B0D8F3183000_ntdll_dl_
- local_ntdll_dll_59B0D8F3183000_ntdll_dll
- local_wkernel32_pdb_D347455996F747D6BF43C176B2171E681_wkernel32_pd_
- local_wkernel32_pdb_D347455996F747D6BF43C176B2171E681_wkernel32_pdb
- local_wkernel32_pdb_D347455996F747D6BF43C176B2171E681_wkernel32_src_zip
- local_wkernel32_pdb_D347455996F747D6BF43C176B2171E681_wkernel32_sym
- local_wntdll_pdb_971F98E5CE6041FFB2D7235BBEB345781_wntdll_pd_
- local_wntdll_pdb_971F98E5CE6041FFB2D7235BBEB345781_wntdll_pdb
- local_wntdll_pdb_971F98E5CE6041FFB2D7235BBEB345781_wntdll_src_zip
- local_wntdll_pdb_971F98E5CE6041FFB2D7235BBEB345781_wntdll_sym

